src/Databases/DatabaseMemory.cpp

```diff
@@ -21,12 +22,78 @@ namespace ErrorCodes
     extern const int UNKNOWN_TABLE;
     extern const int LOGICAL_ERROR;
     extern const int INCONSISTENT_METADATA_FOR_BACKUP;
+    extern const int INCORRECT_QUERY;
 }
 
-DatabaseMemory::DatabaseMemory(const String & name_, ContextPtr context_)
+class MemoryDatabaseTablesSnapshotIterator final : public DatabaseTablesSnapshotIterator
+{
+public:
+    explicit MemoryDatabaseTablesSnapshotIterator(DatabaseTablesSnapshotIterator && base)
+        : DatabaseTablesSnapshotIterator(std::move(base)) {}
+    UUID uuid() const override { return table()->getStorageID().uuid; }
+};
+
+DatabaseMemory::DatabaseMemory(const String & name_, UUID uuid, ContextPtr context_)
     : DatabaseWithOwnTablesBase(name_, "DatabaseMemory(" + name_ + ")", context_)
-    , data_path("data/" + escapeForFileName(database_name) + "/")
-{}
+    , data_path("store/")
+    , path_to_table_symlinks(fs::path(getContext()->getPath()) / "data" / escapeForFileName(name_) / "")
+    , db_uuid(uuid)
+{
+    assert(db_uuid != UUIDHelpers::Nil);
+    fs::create_directories(path_to_table_symlinks);
+}
+
+String DatabaseMemory::getTableDataPath(const String & table_name) const
+{
+    std::lock_guard lock(mutex);
+    auto it = table_name_to_path.find(table_name);
+    if (it == table_name_to_path.end())
+        throw Exception(ErrorCodes::UNKNOWN_TABLE, "Table {} not found in database {}", table_name, database_name);
+    assert(it->second != data_path && !it->second.empty());
+    return it->second;
+}
+
+String DatabaseMemory::getTableDataPath(const ASTCreateQuery & query) const
+{
+    auto tmp = data_path + DatabaseCatalog::getPathForUUID(query.uuid);
+    assert(tmp != data_path && !tmp.empty());
+    return tmp;
+}
+
+void DatabaseMemory::tryCreateSymlink(const String & table_name, const String & actual_data_path, bool if_data_path_exist)
```

Comment:
Consider moving these methods to a common base of `DatabaseMemory` and `DatabaseAtomic`